### PR TITLE
Introduces the ZoomToExtent Button

### PR DESCRIPTION
**Pull request by amir-ba**
### Description:
Introduces the `ZoomToExtent` Button

### CODE REVIEW

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    ]).isRequired
+  }
+
+  /**
+   * Called when the button is clicked.
+   *
+   * @method
+   */
+  onClick = () => {
+    const{map,extent} = this.props;
```
Comment (annarieger) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx]: missed whitespace

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.example.jsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }} />
+
+    <div className="example-block">
+      <span>Zoom to extent button:</span>
+      <ZoomToExtentButton map={map} extent={
+        [588947.9928934451,
+        6584461.475575979,
+        1053685.1248673166,
+        6829059.966088544]
+      }>
```
Comment (annarieger) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.example.jsx]: indentation looks a bit strange

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import OlMap from 'ol/map';
+import OlExtent from 'ol/extent';
+import OlSimpleGeometry from 'ol/geom/simplegeometry';
+import {
+  SimpleButton
+} from '../../index';
+import { TestUtil } from '../../Util/TestUtil';
```
Comment (annarieger) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx]: Do you really need the `TestUtil` class here? It will be mostly used only in test classes..

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.spec.jsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+describe('<ZoomToExtentButton />', () => {
+
+  let map;
+  let mockGeomery;
+  let mockExtent;
+
+  beforeEach(() => {
+    map = TestUtil.createMap();
+    mockExtent = TestUtil.generateExtent()
```
Comment (annarieger) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.spec.jsx]: missing semicolon

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.spec.jsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    const wrapper = TestUtil.mountComponent(ZoomToExtentButton, {
+      map,
+      extent: mockExtent
+    });
+    wrapper.instance().onClick();
+    const newExtent = map.getView().calculateExtent();
+    const newSize = OlExtent.getSize(newExtent)
+    const mockSize = OlExtent.getSize(mockExtent)
+    const newCenter=OlExtent.getCenter(newExtent)
+    const mockCenter=OlExtent.getCenter(mockExtent)
```
Comment (annarieger) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.spec.jsx]: missing semicolons

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx (hunk 3 above)
Comment (KaiVolland) [reply src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx]: :+1: Plz remove

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+     */
+    map: PropTypes.instanceOf(OlMap).isRequired,
+    
+    /**
+     * The array extent[minx, miny, maxx, maxy]  (the values must be in the maps coordination system)
+     * or instance of Ol SimpleGeometry that the map should zoom to.
+     * 
+     */
+    extent: PropTypes.oneOfType([
+      PropTypes.array,//.instanceOf(OlExtent),
```
Comment (KaiVolland) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx]: You could enhance this with `PropTypes.arrayOf(PropTypes.number)`

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * Called when the button is clicked.
+   *
+   * @method
+   */
+  onClick = () => {
+    const{map,extent} = this.props;
+    const view = map.getView();
+    view.fit(extent,{ constrainResolution: false});
```
Comment (KaiVolland) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx]: Would be cool if you add `fitOptions` as a prop to the class. You could then replace this `{ constrainResolution: false}` with the prop.

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  static propTypes = {
+    /**
+     * The className which should be added.
+     * @type {String}
+     */
+    className: PropTypes.string,
+
+    /**
+     * Instance of OL map this component is bound to.
+     *
```
Comment (dnlkoch) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx]: Missing JSDoc.

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    /**
+     * Instance of OL map this component is bound to.
+     * 
+     */
+    map: PropTypes.instanceOf(OlMap).isRequired,
+    
+    /**
+     * The array extent[minx, miny, maxx, maxy]  (the values must be in the maps coordination system)
+     * or instance of Ol SimpleGeometry that the map should zoom to.
+     *
```
Comment (dnlkoch) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx]: Missing JSDoc.

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.spec.jsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+  beforeEach(() => {
+    map = TestUtil.createMap();
+    mockExtent = TestUtil.generateExtent()
+    mockGeomery = TestUtil.generatePolygonGeometry();
+  });
+
+
+
+
```
Comment (dnlkoch) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.spec.jsx]: Please remove empty lines.

**Comment by amir-ba** [comment]
Thanks for the comments I fixed all of them.

File: .vscode/launch.json [hunk 11]
```diff
@@ -1,17 +0,0 @@
...
-        "runtimeArgs": [
-          "--inspect-brk",
-          "${workspaceRoot}/node_modules/.bin/jest",
-          "--runInBand"
-        ],
-        "console": "integratedTerminal",
-        "internalConsoleOptions": "neverOpen"
-      }
-    ]
-  }
```
Comment (marcjansen) [inline .vscode/launch.json]: why is this deleted?

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * The className which should be added.
+     * @type {String}
+     */
+    className: PropTypes.string,
+
+    /**
+     * Instance of OL map this component is bound to.
+     * @type {OLMAP}
+     */
```
Comment (marcjansen) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx]: Type seems wrong

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.spec.jsx [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    });
+    wrapper.instance().onClick();
+    const newExtent = map.getView().calculateExtent();
+    const newSize = OlExtent.getSize(newExtent);
+    const mockSize = OlExtent.getSize(mockExtent);
+    const newCenter=OlExtent.getCenter(newExtent);
+    const mockCenter=OlExtent.getCenter(mockExtent);
+    expect(newCenter).toEqual(mockCenter);
+    if (!(mockSize[1] <newSize[1]))     expect(newSize[1]).toEqual(mockSize[1]);
+    if (!(mockSize[0] <newSize[0]))     expect(newSize[0]).toEqual(mockSize[0]);
```
Comment (marcjansen) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.spec.jsx]: whitespace missing here and above

File: src/Util/TestUtil.js [hunk 14]
```diff
@@ -168,6 +170,41 @@ export class TestUtil {
...
+    const extent= OlExtent.boundingExtent(polyCoords);
+    const geom = new OlGeomPolygon.fromExtent(extent); 
+    return geom;
+  })
+
+  /**
+   * Returns a random extent .
+   * @type {array}
+   */
+  static generateExtent = (() => {
```
Comment (marcjansen) [inline src/Util/TestUtil.js]: Unneeded `(` (and later `)` ), I guess

File: src/Util/TestUtil.js [hunk 15]
```diff
@@ -168,6 +170,41 @@ export class TestUtil {
...
+    const polyCoords = [];
+    for (let i =0;i<5;i++){
```
Comment (marcjansen) [inline src/Util/TestUtil.js]: whitespace. we should have a rule enforcing this. can be done in a separate PR

File: src/Util/TestUtil.js [hunk 16]
```diff
@@ -168,6 +170,41 @@ export class TestUtil {
...
+  static generatePolygonGeometry = (() => {
+    const polyCoords = [];
+    for (let i =0;i<5;i++){
+      const coords = [
+        Math.floor(Math.random() * 180) - 180,
+        Math.floor(Math.random() * 90) - 90
+      ];
+      polyCoords.push(coords);
+    }
+    const extent= OlExtent.boundingExtent(polyCoords);
```
Comment (marcjansen) [inline src/Util/TestUtil.js]: whitespace

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.example.md [hunk 17]
```diff
@@ -0,0 +1,8 @@
+---
+layout: basic.hbs
+title: ZoomToExtentButton example
+description: This is a example using an ZoomToExtentButton.
```
Comment (marcjansen) [inline src/Button/ZoomToExtentButton/ZoomToExtentButton.example.md]: an example

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx (hunk 12 above)
Comment (KaiVolland) [reply src/Button/ZoomToExtentButton/ZoomToExtentButton.jsx]: yes plz use `ol.map` (dot notation)

File: src/Button/ZoomToExtentButton/ZoomToExtentButton.spec.jsx (hunk 13 above)
Comment (KaiVolland) [reply src/Button/ZoomToExtentButton/ZoomToExtentButton.spec.jsx]: New Issue: #522

File: src/Util/TestUtil.js (hunk 15 above)
Comment (KaiVolland) [reply src/Util/TestUtil.js]: New Issue: #522

File: .vscode/launch.json (hunk 11 above)
Comment (amir-ba) [reply .vscode/launch.json]: My mistake I though I have added it from my own files so I removed it. I'll just restore it.

**Comment by KaiVolland** [comment]
Great!


When @annarieger gives here ok too you can merge this.

**Comment by annarieger** [comment]
Good to go, thanks!

**Comment by amir-ba** [comment]
I don't think I have write access to the repo to merge my pull request

**Comment by annarieger** [comment]
I've just sent you an invitation to join terrestris repo. Please check your mails.

**Comment by amir-ba** [comment]
Tnx @annarieger
